middleware.go

```
@@ -32,25 +33,41 @@ func newMiddlewareWrapper(handlerInterface interface{}, middlewareChain LambdaFu
 		return errorHandler(err)
 	}
 	takesContext := handlerTakesContext(handlerType)
-
+	var eventType reflect.Type
+	if (!takesContext && handlerType.NumIn() == 1) || handlerType.NumIn() == 2 {
+		eventType = handlerType.In(handlerType.NumIn() - 1)
+	}
 	return func(ctx context.Context, payload []byte) (interface{}, error) {
 		log.Println("[newMiddlewareWrapper] wrapped function handler")
 
 		ctx = context.WithValue(ctx, ctxKeyPayload, payload)
-		var event interface{}
-		if (!takesContext && handlerType.NumIn() == 1) || handlerType.NumIn() == 2 {
-			eventType := handlerType.In(handlerType.NumIn() - 1)
-			ctx = context.WithValue(ctx, ctxKeyTIn, eventType)
-			evt := reflect.New(eventType)
-
-			if err := json.Unmarshal(payload, evt.Interface()); err != nil {
-				return nil, err
-			}
-			event = evt.Elem().Interface()
+		if eventType == nil {
+			return middlewareChain(ctx, payload)
 		}
+		ctx = context.WithValue(ctx, ctxKeyTIn, eventType)
+
+		// If we try to unmarshal here and TIn is an array type, this will fail
+		// as the JSON unmarshaller will expect a '[' char and gets a '{' instead.
+		// So scenarios are:
+		// TIn = events.SQSEvent - Unmarshalling here works
+		// TIn = []interface{} - Unmarshalling here fails... what do we do :(
+		// TIn = interface{} - Unmarshalling here works
+		//
+		//evt, err := unmarshalToType(json.Unmarshal, eventType, payload)
```

So yes, I think being able to enable/disable this would be ideal, because in the simplest cases having to create a middleware to unmarshal makes for a poor experience.

It's really the list option that causes issues, because JSON events that aren't arrays just silently fail (but I don't think that's a nice behaviour to rely on).

A couple of options we could consider for this function:

1. Paramaterise this function signature to enable/disable it (e.g. useDefaultUnmarshaller)
1. Parameterise it to accept a default unmarshaller (and if the type is `nil` we skip unmarshalling)